### DISCUSSION TITLE
Trust absolute timestamps in linux audio/video sources

### DISCUSSION THREAD

**Discussion by viric** [obsproject/obs-studio#11108] [Development Questions]
OBS is structured towards synchronising audio/video based on *relative* timestamps, instead of absolute timestamps.
For example, linux-v4l2 changes all frame timestamps relative to first frame: https://github.com/obsproject/obs-studio/blob/1451554fb8f05d9a92452b5d8bd4dfa18fb56578/plugins/linux-v4l2/v4l2-input.c#L267-L270
Also, linux-pulseaudio ignores the timestamp set in pulseaudio frames, and picks the cpu timestamp at frame reception time, with a hardcoded offset: https://github.com/obsproject/obs-studio/blob/1451554fb8f05d9a92452b5d8bd4dfa18fb56578/plugins/linux-pulseaudio/pulse-input.c#L212-L215

The sequential startup of drivers makes all relative timestamps to be offset between them. For example I have to delay about 200ms my audio, to match video. Then, due to timestamps at reception time, OBS stills needs to figure out how to synchronise all them without common clock (https://github.com/obsproject/obs-studio/blob/1451554fb8f05d9a92452b5d8bd4dfa18fb56578/libobs/obs-source.c#L1556-L1571 etc). That's a hard problem and we often see how streams get AV-desync because of eventual busy CPU.

But the fact is that the both v4l2 and pulseaudio timestamp audio and video frames with good *absolute CPU timestamps* set at interrupt time, that could be perfectly used for proper av sync without any effort 

I think that it would be nice if the obs_source_frame allowed a flag that said "timestamp is in absolute time" (in os_gettime_ns() clock), so drivers can provide timestamps in absolute time. That would solve many AV-desync problems and it's trivial to implement I think. At least for v4l2+pulseaudio.

Would any of that be welcome?

**Comment by kkartaltepe** [comment]
Sure but you would need to verify it is impossible to send invalid timestamps, as such all timestamps must not ever originate in devices (or untrusted code) in any case. I don't think this is actually true and that v4l2 and pulseaudio probably have options to trust devices. If that is the case then we must continue not to trust timestamps which may be garbage but just so happen not to be garbage for you.

**Comment by viric** [comment]
What do you mean about untrusted code? In v4l2 and pulseaudio I think you can ask in what clock the timestamps are. You would have to make sure it's in the same clock as the os_gettime_ns(), which is the usual monotone clock afaik. That's what the linux drivers will usually provide, additional to more timestamps if required.
Some webcams, for example a Creative webcam I have, has an internal clock. And the driver can be asked to figure out the offset between the two clocks, so you can have cpu-monotone-clock timestamps at actual exposure moment. Not all webcams/drivers support that though, or by default.
In case the timestamp comes from the interrupt code, instead of the exposure moment, that's nevertheless a lot more stable than anything you can get otherwise.
I don't see much difference into trusting the absolute timestamps or the relative timestamps. They are taken either with the same clock or not.

**Reply by kkartaltepe** [reply]
> What do you mean about untrusted code?

Literally any code you cannot trust. I think we can trust that if we see kernel code adding timestamps via ktime its right. If you expect us to trust random device firmware or userspace programs its not going to happen.

> I think you can ask in what clock the timestamps are.

I think this is definitely required if timestamps other than those from the v4l2 or pulseaudio system code are used. But its not sufficient. We need to know those clock source reports can be trusted. The second any poorly authored or cleverly authored code can report the wrong clock source it just introduces a new way for desync to enter, and we should just stick with the existing code.

**Reply by stephematician** [reply]
Aren't the situations already where timestamps are 'trusted' as being system timestamps - e.g. 381afe0 introduces a test that timestamps are within 2s of the system clock, and addresses the following problem:

> For example, if the first set of audio samples from the audio source came in delayed, while the subsequent samples were not delayed, those first samples could have potentially inadvertently triggered the timing adjustments, which would affect all subsequent audio samples.

As OP noted; _pulseaudio_ sources provide timestamps based on the system clock (and offset by 0.5sec according to the value of `STARTUP_TIMEOUT_NS`); so they'll still pass the "within 2s" test and `timing_adjust` is set to zero as a result.

There doesn't appear to be much difference between _v4l2-input_ sources whether they send a _relative_ timestamp or the timestamp as reported by V4L2 - except that doing the latter avoids a (obvious) bug in `async_tick` where both the first and second frames received will be immediately dequeued (because `source->last_frame_ts` is zero in both cases). I believe the intention is that only the _first_ frame is immediately dequeued.

**Comment by viric** [comment]
The only timestamps we can trust have to be in kernel monotonic clock, sure. Any other clock would require synchronisation.
The webcam driver can, if the user desires, use the webcam clock but it will care about translating it to kernel monotonic clock. That's up to driver parameters, not the application.

**Comment by viric** [comment]
Hm maybe it's not so easy to get the monotonic kernel timestamps from the pulseaudio driver. The alsa driver provides them, but not pulseaudio.

I'm checking in my system with linux-v4l2 and linux-pulseaudio, and it seems that for v4l2 I have a timestamp-to-systemtime of 32ms, and for pulseaudio a latency between 10 and 30ms.

I'm struggling to understand why I need to set a 200ms sync_offset to the audio stream, to get more or less synchronous AV in obs.

**Reply by boyokoherdy-oss** [reply]
cool

**Comment by viric** [comment]
It seems I had enabled "Low latency audio" in settings; disabling this, pulseaudio recording latency went down from the ~10-30ms to quite steady 9ms.
